### PR TITLE
Fixes #2989 Timeout performing ViewQuery

### LINKED ISSUE #2989: Workshop error - Unexpected FATAL: Error opening database: Couldn't create user "...": Timeout performing ViewQuery - could indicate that views are still reindexing
I'm working through the Connect Workshop instructions, and ran into this on startup of the Sync Gateway docker container:

```
2017-10-23T18:32:49.222Z Import+: Imported landmark_25709 (delete=false) as rev 1-c7d90a549f551182f72ec252cc05c351
2017-10-23T18:32:49.222Z Import+: Attempting to import doc "route_21406"...
2017-10-23T18:32:49.222Z CRUD+: Doc route_21406 is not an SG write, based on cas. cas:14f045afb8260000 syncCas:""
2017-10-23T18:32:49.223Z WARNING: channelComputer.ComputeChannelsForPrincipal returned error for &{{demo   0  <nil>} { false [] <nil>   []} 0xc420296320 []}: Timeout performing ViewQuery - could indicate that views are still reindexing -- auth.(*Authenticator).rebuildChannels() at auth.go:156
2017-10-23T18:32:49.223Z FATAL: Error opening database: Couldn't create user "demo": Timeout performing ViewQuery - could indicate that views are still reindexing -- rest.RunServer() at config.go:769
```

## SG config:

https://gist.github.com/tleyden/365f621ffa3f428fd8c47ad8840caaf3

## Actual vs expected:  

If the view query times out in this case, it feels like it should do a backoff retry

### PR DESCRIPTION
**Pull request by tleyden**
Fixes https://github.com/couchbase/sync_gateway/issues/2989

## SG Accel dependent PR

- [x] https://github.com/couchbaselabs/sync-gateway-accel/pull/191

## Integration tests

http://uberjenkins.sc.couchbase.com/view/Build/job/sync-gateway-integration-master/248/
http://uberjenkins.sc.couchbase.com/view/Build/job/sync-gateway-integration-master/249/

### CODE REVIEW

File: rest/server_context.go [hunk 1]
```diff
@@ -815,21 +815,44 @@ func (sc *ServerContext) installPrincipals(context *db.DatabaseContext, spec map
...
+				return true, err, nil
 			}
-		} else if isGuest {
+
+			// No errors, assume it worked
+			return false, nil, nil
+
+		}
+
+		err, _ := base.RetryLoop("installPrincipals", worker, base.CreateDoublingSleeperFunc(16, 10))
```
Comment (tleyden) [inline rest/server_context.go:843]: Wasn't sure what to use here.

File: rest/server_context.go [hunk 2]
```diff
@@ -815,21 +815,44 @@ func (sc *ServerContext) installPrincipals(context *db.DatabaseContext, spec map
...
+
+		worker := func() (shouldRetry bool, err error, value interface{}) {
+			_, err = context.UpdatePrincipal(*princ, (what == "user"), isGuest)
+			if err != nil {
+				if base.IsConflictError(err) {
+					// Ignore and absorb this error if it's a conflict error, which just means that updatePrincipal didn't overwrite an existing user.
+					// Since if there's an existing user it's "mission accomplished", this can be treated as a success case.
+					return false, nil, nil
+				}
+				// Assume it might be a timeout error due to view re-indexing.  So retry
```
Comment (tleyden) [inline rest/server_context.go]: Should this be comparing against `ErrViewTimeoutError` and only retry in that case?

Comment (adamcfraser) [reply rest/server_context.go]: Yeah - I think we should be handling the timeout error specifically, and not just retrying on everything except CAS.

File: rest/server_context.go [hunk 3]
```diff
@@ -815,21 +815,44 @@ func (sc *ServerContext) installPrincipals(context *db.DatabaseContext, spec map
...
-		_, err := context.UpdatePrincipal(*princ, (what == "user"), isGuest)
-		if err != nil {
-			// A conflict error just means updatePrincipal didn't overwrite an existing user.
-			if status, _ := base.ErrorAsHTTPStatus(err); status != http.StatusConflict {
-				return err
+
+		worker := func() (shouldRetry bool, err error, value interface{}) {
+			_, err = context.UpdatePrincipal(*princ, (what == "user"), isGuest)
+			if err != nil {
+				if base.IsConflictError(err) {
```
Comment (adamcfraser) [inline rest/server_context.go]: Can we use the existing IsCasMismatch instead of IsConflictError?  They seem redundant.

**Review by adamcfraser — CHANGES_REQUESTED** [review]

File: rest/server_context.go (hunk 2 above)
Comment (tleyden) [reply rest/server_context.go]: Fixed

File: rest/server_context.go (hunk 3 above)
Comment (tleyden) [reply rest/server_context.go]: Fixed.  Also I removed un unused param from `IsCasMismatch()`, which requires a companion change to SG Accel.

File: rest/server_context.go (hunk 1 above)
Comment (tleyden) [reply rest/server_context.go:843]: @adamcfraser does this look like a reasonable retry value?

`err, _ := base.RetryLoop("installPrincipals", worker, base.CreateDoublingSleeperFunc(16, 10))`

Or would it make more sense to switch away from the backoff retry and just do a normal retry?

**Comment by tleyden** [comment]
## Integration tests re-run in progress

http://uberjenkins.sc.couchbase.com/view/Build/job/sync-gateway-integration-master/240/
http://uberjenkins.sc.couchbase.com/view/Build/job/sync-gateway-integration-master/241/

**Comment by tleyden** [comment]
One of the integration test runs failed w/ what looks like a temporary error.  I filed a separate ticket for that: https://github.com/couchbase/sync_gateway/issues/3134

**Comment by tleyden** [comment]
Another integration test temp failure: https://github.com/couchbase/sync_gateway/issues/3006.  Re-running.

**Review by adamcfraser — APPROVED** [review]

**Comment by adamcfraser** [comment]
Looks good - merged accel PR and bumped manifest.  Will merge once drone build completes.

**Pull request merged**